### PR TITLE
do_patch: Add local patches capability. Closes #1176 #1178

### PR DESCRIPTION
**Pull request by 1480c1**
@lars18th Please check if this is okay.

### CODE REVIEW

**Comment by lars18th** [comment]
Hi @1480c1 ,

It works!
However, I have one question:
- If I use `patch -p1 -i "$LOCALBUILDDIR/ffmpeg-001-my-patch.patch` in `ffmpeg_extra.sh` it works.
- But if I try `do_patch "$LOCALBUILDDIR/ffmpeg-001-my-patch.patch"` or `do_patch "./ffmpeg-001-my-patch.patch"` or `do_patch "ffmpeg-001-my-patch.patch"` it fails. Why?

In any case, please merge it as it works (and close the other superseded PR).

**Comment by 1480c1** [comment]
does it show any message when it fails?

File: build/media-suite_helper.sh [hunk 1]
```diff
@@ -1031,8 +1031,9 @@ do_patch() {
         do_wget -c -r -q "$patch" "$patchName"
     elif [[ -f "$patch" ]]; then
         # Check if the patch is a local patch and copy it to the current dir
+        patch="$(realpath $patch)" # Resolve fullpatch
```
Comment (wiiaboo) [inline build/media-suite_helper.sh]: Might wanna quote-surround `$patch`, or does `realpath` work with spaces?

Comment (1480c1) [reply build/media-suite_helper.sh]: Honestly, I'm not exactly sure.

File: build/media-suite_helper.sh [hunk 2]
```diff
@@ -1010,32 +1011,52 @@ do_removeOptions() {
...
+    [[ "$PWD" = "$LOCALBUILDDIR" ]] &&
+        do_exit_prompt "Running patches in the build folder is not supported.
+        Please make a patch for individual folders or modify the script directly"
+
+    if [[ ${patch:0:4} = "http" ]] || [[ ${patch:0:3} = "ftp" ]]; then
+        # Filter out patches that would require curl
+        do_wget -c -r -q "$patch" "$patchName"
+    elif [[ -f "$patch" ]]; then
+        # Check if the patch is a local patch and copy it to the current dir
+        patch="$(realpath $patch)" # Resolve fullpatch
```
Comment (1480c1) [inline build/media-suite_helper.sh]:
```suggestion
        patch="$(realpath "$patch")" # Resolve fullpatch
```

**Comment by lars18th** [comment]
> does it show any message when it fails?

- In the main window: NO, nothing is printed.
- Inside the `ab-pre_configure.log` file: YES. The same error about "I can't download file" or it ask about user interaction.

**Comment by lars18th** [comment]
Hi @1480c1 ,

I checked last version with the quotes around the patch variable: OK.
Now `do_patch "$LOCALBUILDDIR/ffmpeg-0001-my-patch.patch"` inside `ffmpeg_extra.sh` works.
Thank you!

However, one advise: using `do_patch` nothing is printed in `ab-suite.pre_configure.log`. With `patch -p1 ...` I see the results. But with `do_patch` it's transparant. Only FYI.

I feel now it's ready to merge. :+1:

**Comment by wiiaboo** [comment]
That's because `patch`/`git am` stdout/err are silenced, not logged or shown ever.

**Comment by lars18th** [comment]
> That's because `patch`/`git am` stdout/err are silenced, not logged or shown ever.
> But it'd make sense to wrap the whole output of `do_patch` in a log, maybe.

OK. As you prefer. For me the PR is ready to merge it!

**Comment by wiiaboo** [comment]
There's usually no need to log the patches. It either works or it doesn't.

**Comment by lars18th** [comment]
> There's usually no need to log the patches. It either works or it doesn't.

OK. So, please merge it.

**Pull request merged**